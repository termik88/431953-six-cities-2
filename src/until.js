export const getCitiesList = (placesAll) => [...new Set(placesAll.map((place) => place.city.name))];

export const getRandomCity = (citiesList) => citiesList[Math.floor(Math.random() * Math.floor(citiesList.length))];

const RATING_STARS = 5;
const PERCENT = 100;
const transformRatingToPercent = (ratingPlace) => Math.round(ratingPlace) / RATING_STARS * PERCENT;

export const transformRatingToStarsNumber = (ratingPlace) => ratingPlace * RATING_STARS / PERCENT;

export const prepareUser = (user) => {
  const userData = Object.assign({}, user);
  userData.avatarUrl = userData.avatar_url;
  userData.isPro = userData.is_pro;
  delete userData.avatar_url;
  delete userData.is_pro;

  return userData;
};

export const preparePlace = (places) => {
  const newPlaces = Object.assign({}, places);
  newPlaces.previewImage = newPlaces.preview_image;
  newPlaces.isFavorite = newPlaces.is_favorite;
  newPlaces.isPremium = newPlaces.is_premium;
  newPlaces.maxAdults = newPlaces.max_adults;
  newPlaces.rating = transformRatingToPercent(newPlaces.rating);
  newPlaces.type = newPlaces.type[0].toUpperCase() + newPlaces.type.slice(1);
  delete newPlaces.preview_image;
  delete newPlaces.is_favorite;
  delete newPlaces.is_premium;
  delete newPlaces.max_adults;
  newPlaces.host = prepareUser(newPlaces.host);

  return newPlaces;
};

export const preparePlaces = (placesList) => placesList.map((place) => preparePlace(place));

export const isValidationEmail = (email) => {
  const pattern = /^([A-Za-z0-9_\-\.])+\@([A-Za-z0-9_\-\.])+\.([A-Za-z]{2,4})$/;
  return pattern.test(String(email).toLowerCase());
};

const MONTHS = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`,
];

export const getMonthYear = (date) => {
  if (!date) {
    return ``;
  }
  const fullDate = new Date(date);
  const month = MONTHS[fullDate.getMonth()];

  return `${month} ${fullDate.getFullYear()}`;
};

const prepareComment = (comment) => {
  const newComment = Object.assign({}, comment);
  newComment.rating = transformRatingToPercent(newComment.rating);
  newComment.commentText = newComment.comment;
  delete newComment.comment;
  newComment.user = prepareUser(newComment.user);

  return newComment;
};

export const prepareComments = (comments) => comments.map((comment) => prepareComment(comment));

export const getPlacesNearest = (currentPoint, allPoints) => {
  const distance = (x1, y1, x2, y2) => Math.sqrt(Math.pow(x2 - x1, 2) + (Math.pow(y2 - y1, 2)));
  const filteredPoints = allPoints.filter((item) => item.city.name === currentPoint.city.name && item.id !== currentPoint.id);
  return filteredPoints
    .map((x) => x)
    .sort((x, y) =>
      Math.abs(distance(x.location.longitude, x.location.latitude, currentPoint.location.longitude, currentPoint.location.latitude))
      - Math.abs(distance(y.location.longitude, y.location.latitude, currentPoint.location.longitude, currentPoint.location.latitude)))
    .slice(0, 3);
};

