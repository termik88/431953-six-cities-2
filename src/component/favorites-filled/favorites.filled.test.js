import React from 'react';
import renderer from 'react-test-renderer';
import {Router} from "react-router";
import {createBrowserHistory} from "history";

import FavoritesFilled from "./favorites-filled.jsx";
import {getCitiesList} from "../../until";
import places from "../../mocks/places";

jest.mock(`../favorite-button/favorite-button.jsx`);

describe(`FavoritesCity component test.`, () => {
  it(`Correctly renders after relaunch`, () => {
    const citiesList = getCitiesList(places);

    const mockDate = {
      favoritesCityList: citiesList,
      favoritesOffers: places,
      handleClickCityName: jest.fn()
    };

    const test = renderer
      .create(
          <Router history={createBrowserHistory()}>
            <FavoritesFilled
              {...mockDate}
            />
          </Router>).toJSON();

    expect(test).toMatchSnapshot();
  });
});
