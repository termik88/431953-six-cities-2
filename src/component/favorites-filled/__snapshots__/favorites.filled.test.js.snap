// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesCity component test. Correctly renders after relaunch 1`] = `
<main
  className="page__main page__main--favorites"
>
  <div
    className="page__favorites-container container"
  >
    <section
      className="favorites"
    >
      <h1
        className="favorites__title"
      >
        Saved listing
      </h1>
      <ul
        className="favorites__list"
      >
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link locations__item-link--test"
                href="/"
                onClick={[Function]}
              >
                <span>
                  Dusseldorf
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            <article
              className="place-card place-card--favorites"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__image-wrapper place-card__image-wrapper--favorites"
              >
                <a
                  href="/offer/1"
                  onClick={[Function]}
                >
                  <img
                    alt="Canal View Prinsengracht"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/9.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info place-card__info--favorites"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      101
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "3.1%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/1"
                    onClick={[Function]}
                  >
                    Canal View Prinsengracht
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  room
                </p>
              </div>
            </article>
            
            
            
            
            
            
            
            
            
          </div>
        </li>
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link locations__item-link--test"
                href="/"
                onClick={[Function]}
              >
                <span>
                  Amsterdam
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            
            <article
              className="place-card place-card--favorites"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__mark"
              >
                <span>
                  Premium
                </span>
              </div>
              <div
                className="place-card__image-wrapper place-card__image-wrapper--favorites"
              >
                <a
                  href="/offer/2"
                  onClick={[Function]}
                >
                  <img
                    alt="House in countryside"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/16.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info place-card__info--favorites"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      163
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "4.5%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/2"
                    onClick={[Function]}
                  >
                    House in countryside
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  room
                </p>
              </div>
            </article>
            
            
            <article
              className="place-card place-card--favorites"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__mark"
              >
                <span>
                  Premium
                </span>
              </div>
              <div
                className="place-card__image-wrapper place-card__image-wrapper--favorites"
              >
                <a
                  href="/offer/5"
                  onClick={[Function]}
                >
                  <img
                    alt="Beautiful & luxurious apartment at great location"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/2.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info place-card__info--favorites"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      463
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "4.9%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/5"
                    onClick={[Function]}
                  >
                    Beautiful & luxurious apartment at great location
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  hotel
                </p>
              </div>
            </article>
            
            
            
            
            
          </div>
        </li>
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link locations__item-link--test"
                href="/"
                onClick={[Function]}
              >
                <span>
                  Cologne
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            
            
            <article
              className="place-card place-card--favorites"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__image-wrapper place-card__image-wrapper--favorites"
              >
                <a
                  href="/offer/3"
                  onClick={[Function]}
                >
                  <img
                    alt="Beautiful & luxurious apartment at great location"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/9.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info place-card__info--favorites"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      254
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "2.1%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/3"
                    onClick={[Function]}
                  >
                    Beautiful & luxurious apartment at great location
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  apartment
                </p>
              </div>
            </article>
            
            
            
            
            
            
            
          </div>
        </li>
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link locations__item-link--test"
                href="/"
                onClick={[Function]}
              >
                <span>
                  Paris
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            
            
            
            <article
              className="place-card place-card--favorites"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__image-wrapper place-card__image-wrapper--favorites"
              >
                <a
                  href="/offer/4"
                  onClick={[Function]}
                >
                  <img
                    alt="The Joshua Tree House"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/1.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info place-card__info--favorites"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      284
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "4.3%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/4"
                    onClick={[Function]}
                  >
                    The Joshua Tree House
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  room
                </p>
              </div>
            </article>
            
            
            
            
            
            
          </div>
        </li>
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link locations__item-link--test"
                href="/"
                onClick={[Function]}
              >
                <span>
                  Brussels
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            
            
            
            
            
            <article
              className="place-card place-card--favorites"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__image-wrapper place-card__image-wrapper--favorites"
              >
                <a
                  href="/offer/6"
                  onClick={[Function]}
                >
                  <img
                    alt="Tile House"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/2.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info place-card__info--favorites"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      241
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "2.9%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/6"
                    onClick={[Function]}
                  >
                    Tile House
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  room
                </p>
              </div>
            </article>
            
            
            
            
          </div>
        </li>
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link locations__item-link--test"
                href="/"
                onClick={[Function]}
              >
                <span>
                  Hamburg
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            
            
            
            
            
            
            <article
              className="place-card place-card--favorites"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__image-wrapper place-card__image-wrapper--favorites"
              >
                <a
                  href="/offer/7"
                  onClick={[Function]}
                >
                  <img
                    alt="Penthouse, 4-5 rooms + 5 balconies"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/18.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info place-card__info--favorites"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      194
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "4.4%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/7"
                    onClick={[Function]}
                  >
                    Penthouse, 4-5 rooms + 5 balconies
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  apartment
                </p>
              </div>
            </article>
            <article
              className="place-card place-card--favorites"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__image-wrapper place-card__image-wrapper--favorites"
              >
                <a
                  href="/offer/8"
                  onClick={[Function]}
                >
                  <img
                    alt="Loft Studio in the Central Area"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/7.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info place-card__info--favorites"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      215
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "3.5%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/8"
                    onClick={[Function]}
                  >
                    Loft Studio in the Central Area
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  room
                </p>
              </div>
            </article>
            <article
              className="place-card place-card--favorites"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__image-wrapper place-card__image-wrapper--favorites"
              >
                <a
                  href="/offer/9"
                  onClick={[Function]}
                >
                  <img
                    alt="The Joshua Tree House"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/18.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info place-card__info--favorites"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      519
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "3.3%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/9"
                    onClick={[Function]}
                  >
                    The Joshua Tree House
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  house
                </p>
              </div>
            </article>
            <article
              className="place-card place-card--favorites"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__image-wrapper place-card__image-wrapper--favorites"
              >
                <a
                  href="/offer/10"
                  onClick={[Function]}
                >
                  <img
                    alt="Beautiful & luxurious apartment at great location"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/7.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info place-card__info--favorites"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      273
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "2%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/10"
                    onClick={[Function]}
                  >
                    Beautiful & luxurious apartment at great location
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  room
                </p>
              </div>
            </article>
          </div>
        </li>
      </ul>
    </section>
  </div>
</main>
`;
